snapshot.go

```diff
@@ -0,0 +1,359 @@
+package frostdb
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/apache/arrow/go/v10/arrow/ipc"
+	"github.com/go-kit/log/level"
+	"github.com/google/btree"
+	"github.com/segmentio/parquet-go"
+
+	"github.com/polarsignals/frostdb/dynparquet"
+	schemav2pb "github.com/polarsignals/frostdb/gen/proto/go/frostdb/schema/v1alpha2"
+	snapshotpb "github.com/polarsignals/frostdb/gen/proto/go/frostdb/snapshot/v1alpha1"
+	"github.com/polarsignals/frostdb/parts"
+)
+
+const (
+	snapshotMagic = "FDBS"
+	dirPerms      = os.FileMode(0755)
+	filePerms     = os.FileMode(0640)
+)
+
+// segmentName returns a 20-byte textual representation of a snapshot file name
+// at a given txn used for lexical ordering.
+func snapshotFileName(tx uint64) string {
+	return fmt.Sprintf("%020d.fdbs", tx)
+}
+
+// snapshot takes a snapshot of the state of the database at transaction tx.
+func (db *DB) snapshot(ctx context.Context, tx uint64) error {
+	if !db.columnStore.enableWAL {
+		// TODO(asubiotto): Should we allow snapshots to be enabled
+		// independently?
+		return nil
+	}
+
+	snapshotsDir := db.snapshotsDir()
+	if err := os.MkdirAll(snapshotsDir, dirPerms); err != nil {
+		return err
+	}
+
+	fileName := filepath.Join(snapshotsDir, snapshotFileName(tx))
+	f, err := os.OpenFile(fileName, os.O_CREATE|os.O_RDWR|os.O_TRUNC, filePerms)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	if err := func() error {
+		if err := writeSnapshot(ctx, tx, db, f); err != nil {
+			return err
+		}
+		return f.Sync()
+	}(); err != nil {
```

Comment:
Interesting, did you use a closure here to encapsulate the code so the context error wrapping can be done once below?

Reply:
Yes, it allows me to centralize cleanup code.